From 728850e73b49b4c84b7d9e10df3d79ee195191c1 Mon Sep 17 00:00:00 2001
From: Yong Zhi <yong.zhi@intel.com>
Date: Tue, 10 Apr 2018 17:16:10 -0500
Subject: [PATCH 1912/2328] media: intel-ipu4: imx319: Fix v4l2-compliance
 G/S_FMT test fails

Message for Open Source:
Fix v4l2-compliance fails on imx319 camera sensor subdev.

Message for Internal:
[Issue/Feature] v4l2-compliance -u /dev/v4l-subdev11

"Control ioctls:"

"fail: v4l2-test-subdevs: fmt.colorspace == ~0U"
"fail: v4l2-test-subdevs: checkMBusFrameFmt(node, fmt.format)"
"test Active VIDIOC_SUBDEV_G/S_FMT: FAIL"

"fail: v4l2-test-controls: subscribe event for control"
"     'Camera Controls' failed Inappropriate ioctl for device"
"test VIDIOC_(UN)SUBSCRIBE_EVENT/DQEVENT: FAIL"

"fail: v4l2-test-controls: failed to find event for control
       'Exposure'"
"test VIDIOC_(UN)SUBSCRIBE_EVENT/DQEVENT: FAIL"

[Root Cause/Changes]
Implementation following example of drivers/media/i2c/ov7740.c
(commit 65b0ae5e3e media: ov7740: Set subdev HAS_EVENT flag)

Change-Id: Ib500acb18d22538faa9c4859463813153363c336
Signed-off-by: Yong Zhi <yong.zhi@intel.com>
Signed-off-by: Meng Wei <wei.meng@intel.com>
---
 drivers/media/i2c/imx319.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/media/i2c/imx319.c b/drivers/media/i2c/imx319.c
index 3d8b574..bb6c086 100644
--- a/drivers/media/i2c/imx319.c
+++ b/drivers/media/i2c/imx319.c
@@ -7,6 +7,7 @@
 #include <linux/pm_runtime.h>
 #include <media/v4l2-ctrls.h>
 #include <media/v4l2-device.h>
+#include <media/v4l2-event.h>
 
 
 #define IMX319_REG_MODE_SELECT		0x0100
@@ -1986,6 +1987,10 @@ static void imx319_update_pad_format(struct imx319 *imx319,
 	fmt->format.height = mode->height;
 	fmt->format.code = imx319_get_format_code(imx319);
 	fmt->format.field = V4L2_FIELD_NONE;
+	fmt->format.colorspace = V4L2_COLORSPACE_DEFAULT;
+	fmt->format.xfer_func = V4L2_XFER_FUNC_DEFAULT;
+	fmt->format.ycbcr_enc = V4L2_YCBCR_ENC_DEFAULT;
+	fmt->format.quantization = V4L2_QUANTIZATION_DEFAULT;
 }
 
 static int imx319_do_get_pad_format(struct imx319 *imx319,
@@ -2243,6 +2248,11 @@ static int imx319_identify_module(struct imx319 *imx319)
 	return 0;
 }
 
+static const struct v4l2_subdev_core_ops imx319_subdev_core_ops = {
+	.subscribe_event = v4l2_ctrl_subdev_subscribe_event,
+	.unsubscribe_event = v4l2_event_subdev_unsubscribe,
+};
+
 static const struct v4l2_subdev_video_ops imx319_video_ops = {
 	.s_stream = imx319_set_stream,
 };
@@ -2255,6 +2265,7 @@ static const struct v4l2_subdev_pad_ops imx319_pad_ops = {
 };
 
 static const struct v4l2_subdev_ops imx319_subdev_ops = {
+	.core = &imx319_subdev_core_ops,
 	.video = &imx319_video_ops,
 	.pad = &imx319_pad_ops,
 };
@@ -2399,7 +2410,8 @@ static int imx319_probe(struct i2c_client *client,
 
 	/* Initialize subdev */
 	imx319->sd.internal_ops = &imx319_internal_ops;
-	imx319->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE;
+	imx319->sd.flags |= V4L2_SUBDEV_FL_HAS_DEVNODE |
+				V4L2_SUBDEV_FL_HAS_EVENTS;
 	imx319->sd.entity.ops = &imx319_subdev_entity_ops;
 	imx319->sd.entity.function = MEDIA_ENT_F_CAM_SENSOR;
 
-- 
2.7.4

